students/jesse_miller/session09/donor_models.py

```diff
@@ -0,0 +1,154 @@
+        self.donations = [] #?  We'll try it.
+        self.count = 0
+        self.total = 0
+        self.average = 0
+
+
+    def donation_add(self, new_donation):
+        '''
+        This should append donations to the list
+        '''
+        self.donations.append(new_donation)
+
+    @property
+    def donation_count(self):
+        '''
+        Fairly obvious, counts the number of donations
+        '''
+        return len(self.donations)
+
+
+    @donation_count.setter
+    def donation_count(self, value):
+        self.count = value
+
+
+    @property
+    def donation_total(self):
+        '''
+        Again, obvious, adds the donations up
+        '''
+        return sum(self.donations)
+
+
+    @donation_total.setter
+    def donation_total(self, value):
+        self.total = value
+
+
+    @property
+    def donation_average(self):
+        '''
+        Averages the donations
+        '''
+        return self.donation_total / self.donation_count
+
+
+    @donation_average.setter
+    def donation_average(self, value):
+        self.average = value
+
+
+################################################################################
+'''
+End Donors
+'''
+################################################################################
+
+class DonorCollection:
+    '''
+    This class will handle the donors dictionary, as well as the collections.
+    Some of the things we can look forward to:
+    Create a donor
+    Find a donor
+    List donors
+    Donor reports
+    Data saving and reloading.
+    '''
+    def __init__(self):
+        '''
+        Initializing and building the default dict
+        '''
+        self.donors_dict = {'Robert Smith': [435.56, 125.23, 357.10],
+                            'JD Cronise': [123.12],
+                            'Chris Stapleton': [243.87, 111.32],
+                            'Dave Lombardo': [63.23, 422.87, 9432.01],
+                            'Randy Blythe': [223.50, 8120.32],
+                            'Devin Townsand': [431.12, 342.92, 5412.45],
+                           }
+
+
+    def donor_creation(self, current_donor):
+        '''
+        Allows for adding new donors to the db
+        '''
+        new_donor = Donor(current_donor)
+        self.donors_dict[current_donor] = new_donor
+        return new_donor
+
+
+    def list_donor(self):
+        '''
+        Listing the contents of the db
+        '''
+        donor_list = []
+        for donor in self.donors_dict:
+            donor_list.append(donor)
+        sorted_donors = sorted(donor_list)
+        return sorted_donors
+
+
+    def find_donor(self, current_donor):
+        '''
+        This will take an inputed donor name, and search for it.  If the name
+        isn't found, it will then bounce the request to the creation function.
+        '''
+        try:
+            return self.donors_dict[current_donor]
+        except KeyError:
+            return self.donor_creation(current_donor)
```

Comment:
Do it the other way around, have _donor_creation_ call _find_donor_ to see if a donor already exists before trying to create a new one.